src/node/miner.cpp

```diff
@@ -394,9 +394,8 @@ void BlockAssembler::addPackageTxs(const CTxMemPool& mempool, int& nPackagesSele
             continue;
         }
 
-        CTxMemPool::setEntries ancestors;
-        std::string dummy;
-        mempool.CalculateMemPoolAncestors(*iter, ancestors, CTxMemPool::Limits::NoLimits(), dummy, false);
+        auto ancestors_result{mempool.CalculateMemPoolAncestors(*iter, CTxMemPool::Limits::NoLimits(), /*fSearchForParents=*/false)};
+        auto ancestors{ancestors_result.value_or(CTxMemPool::setEntries{})};
```

Comment:
Great question - and it pushed me quite deep down an interesting rabbit hole. I hadn't considered this. I think you're right that (without compiler optimization*) this (and all the other similar instances in this PR) would lead to a copy of the set, because we're calling `value_or()` on an lvalue so the [`const&` qualified `value_or()` method](https://github.com/bitcoin/bitcoin/blob/968f03e65c98a9a36aea6c57a29306839b0ada57/src/util/result.h#L60-L64) is used, which returns by value and doesn't benefit from move semantics.

By turning `ancestors_result` into an rvalue ref with `std::move`, we can benefit from the move semantics that `std::set` has implemented since now we'll be using the [`&&` qualified `value_or()` method](https://github.com/bitcoin/bitcoin/blob/968f03e65c98a9a36aea6c57a29306839b0ada57/src/util/result.h#L65-L69).

I've force pushed a change to replace all instances of `ancestors_result` into `std::move(ancestors_result)` as well as update `AssumeCalculateMemPoolAncestors()` to be explicit about using move semantics. _(note: most of these changes are undone in a later commit where we start using the newly introduced `AssumeCalculateMemPoolAncestors()` function)_

*I think, in practice, in most of the affected instances in this PR the copy operation actually would have been optimized away by the compiler (because I think `ancestors_result` just would have been inlined anyway, making them rvalues), but at the very least the code is now more explicit about intent, and potentially more performant - and I could be wrong about my optimization assumptions.

Thank you for raising this!

Reply:
I'm late to this party... If the `T&& value()` member function could be added to the `Result` class, the following code will use a move assignment:
```c++
        auto ancestors_result{mempool.CalculateMemPoolAncestors(*iter, CTxMemPool::Limits::NoLimits(), /*fSearchForParents=*/false)};
        CTxMemPool::setEntries ancestors;
        if (ancestors_result.has_value()) {
            ancestors = ancestors_result.value();
        }
```